waku/v2/protocol/waku_rln_relay/waku_rln_relay_utils.nim

```diff
@@ -1145,19 +1146,39 @@ proc mountRlnRelayDynamic*(node: WakuNode,
   node.wakuRlnRelay = rlnPeer
   return ok(true)
 
-proc readPersistentRlnCredentials*(path: string) : RlnMembershipCredentials {.raises: [Defect, OSError, IOError, Exception].} =
-  info "Rln credentials exist in file"
+proc writeRlnCredentials*(path: string, credentials: RlnMembershipCredentials, password: string) : RlnRelayResult[void] =
+  info "Storing RLN credentials"
+  var jsonString: string
+  jsonString.toUgly(%credentials)
+  let keyfile = createKeyFileJson(toBytes(jsonString), password)
+  if keyfile.isErr():
+    return err("Error while creating keyfile for RLN credentials")
+  if saveKeyFile(path, keyfile.get()).isErr():
+    return err("Error while saving keyfile for RLN credentials")
+  return ok()
+
+proc readRlnCredentials*(path: string, password: string) : RlnRelayResult[Option[RlnMembershipCredentials]] {.raises: [Defect, IOError, Exception].} =
```

Comment:
I don't think this `IOError` should be raised by the `proc` if it's already caught in the body. Also, do we know where the `Exception` comes from?

Reply:
Ah, no. Not like that 😄.

If an error is caught it is _not_ raised by the function anymore (unless explicitly re-raised). In other words, the `raises:` pragma is to list the exceptions that this function is "allowed" to still raise if called elsewhere. If someone calls `readRlnCredentials()` there should be no way that an `IOError` is raised if it's already caught and handled in the func body. If a specific error is not handled in the func body, it should be listed as specifically as possible in the `raises:` list, so that calling functions don't have to implement catch-all blocks (which may obscure serious and unrelated errors).

The changes in https://github.com/status-im/nwaku/commit/01ebc549f5000d3e0021fc63b2c815e1a8165af2 adds a catch-all, which would mean that any unrelated errors will quietly be swallowed.
In other words:
- where reasonable, catch and handle specific errors inside the func body and use the `Result` return type to indicate `error` or `ok` results.
- where it makes sense for an error to go uncaught in a body proc (so that it's raised in the calling function), list it explicitly in the `raises:` list.